ros/src/computing/perception/localization/lib/autoware_lidar_localizer/src/map_manager/map_manager.cpp

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright 2015-2019 Autoware Foundation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "autoware_lidar_localizer/map_manager/map_manager.h"
+
+#include <boost/filesystem.hpp>
+
+template <class PointTarget>
+MapManager<PointTarget>::MapManager()
+    : separate_map_size_(100.0), save_map_leaf_size_(0.2),
+      default_reserve_size_(10000000), merged_map_ptr_(nullptr),
+      is_thread_run_ok_(true), save_added_map_(false) {
+  merged_map_.points.reserve(default_reserve_size_);
+  runProcess();
+}
+
+template <class PointTarget> MapManager<PointTarget>::~MapManager() {
+  is_thread_run_ok_ = false;
+  process_thread_.join();
+}
+
+template <class PointTarget> void MapManager<PointTarget>::runProcess() {
+  process_thread_ = std::thread([this]() {
+    while (is_thread_run_ok_) {
+      if (process_queue_.empty()) {
+        std::this_thread::sleep_for(std::chrono::milliseconds(1));
+        continue;
+      }
+
+      mtx_.lock();
+      auto process = process_queue_.front();
+      process_queue_.pop_front();
+      mtx_.unlock();
+
+      process();
+    }
+  });
+}
+
+template <class PointTarget>
+void MapManager<PointTarget>::addPointCloudMapThread(const boost::shared_ptr<pcl::PointCloud<PointTarget>> &points_raw_ptr) {
```

Comment:
If you create threads like this outside of ROS, we cannot manage execution time properly.

Reply:
@gbiggs So, should I use ros::Timer?